oeps/oep-0003.rst

```diff
@@ -0,0 +1,144 @@
+===================================
+OEP-3: Asynchronous Task Management
+===================================
+
++---------------+-------------------------------------------+
+| OEP           | :doc:`oep-0003`                              |
++---------------+-------------------------------------------+
+| Title         | Asynchronous Task Management              |
++---------------+-------------------------------------------+
+| Last-Modified | 2016-06-13                                |
++---------------+-------------------------------------------+
+| Author        | Jeremy Bowman <jbowman@edx.org>           |
++---------------+-------------------------------------------+
+| Arbiter       |                                           |
++---------------+-------------------------------------------+
+| Status        | Draft                                     |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2016-06-09                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+Proposes a system which allows users to initiate, monitor, and retrieve the
+results of various long-running tasks in a consistent manner.  This system
+leverages existing libraries where practical and makes the development of new
+asynchronous tasks straightforward.
+
+
+Specification
+=============
+
+A new open source Django application will be written to support a user
+interface for managing celery tasks in a consistent manner across services.
+The main goals of the application are, roughly in order:
+
+1. Allow a long-running task triggered by a user to be run asynchronously,
+   while it is clear to the user that they are free to do other work on the
+   site in the meantime
+2. Allow the user to fetch any output of the task after it finishes
+3. Notify the user when the task is complete
+4. Give the user some indication of how far along an in-progress task is
+5. Support prevention of multiple identical tasks from executing concurrently
+
+Celery will be used to run the asynchronous tasks, but the design should be
+capable of adapting to a different task runner later if necessary.  Database
+models will be used to represent tasks as well as their input, output, and
+current status.  It will be possible to present a fan-out of Celery tasks
+instantiated to perform work in parallel as a single task from the user's
+perspective with a meaningful progress indicator.
+
+For notifications, email is sufficient for the initial implementation (but
+other push notifications may be added later).  A task status page for each
```

Comment:
@efagin, @jmbowman @doctoryes @jbarciauskas I was wondering what other people's opinions are on this? What happens when the user has an out of date email address how will they find this download? Also what about emails going in to the spam filter?

I think it would behove us to implement a tracking page on the first iteration.